ui/core/components/suggest_gems_action.ts

```diff
@@ -0,0 +1,627 @@
+
+		// Next, socket and activate the meta
+		optimizedGear = optimizedGear.withMetaGem(this.sim.db.lookupGem(this.metaGemID));
+		optimizedGear = this.activateMetaGem(optimizedGear);
+		await this.updateGear(optimizedGear);
+
+		// Now loop through all gem colors where a priority list has been defined
+		for (var gemColor of GemOptimizer.allGemColors) {
+			if (this.gemPriorityByColor[gemColor].length > 0) {
+				optimizedGear = await this.fillGemsByColor(optimizedGear, gemColor);
+			}
+		}
+
+		// Finally, substitute JC gems by priority while respecting stat caps
+		if (this.useJcGems) {
+			optimizedGear = await this.substituteJcGems(optimizedGear);
+		}
+	}
+	
+	async updateGear(gear: Gear): Promise<Stats> {
+		this.player.setGear(TypedEvent.nextEventID(), gear);
+		await this.sim.updateCharacterStats(TypedEvent.nextEventID());
+		return Stats.fromProto(this.player.getCurrentStats().finalStats);
+	}
+
+	/**
+	 * Helper method for meta gem activation.
+	 *
+	 * @remarks
+	 * Based on the ansatz that most specs are forced to use a suboptimal gem color in
+	 * order to statisfy their meta requirements. As a result, it is helpful to
+	 * compute the item slot in a gear set that provides the strongest socket bonus 
+	 * for that color, since this should minimize the "cost" of activation.
+	 *
+	 * @param gear - Ungemmed gear set
+	 * @param color - Socket color used for meta gem activation
+	 * @param singleOnly - If true, exclude items containing more than one socket of the specified color. If false, instead normalize the socket bonus by the number of such sockets.
+	 * @param blacklistedColor - If non-null, exclude items containing any sockets of this color (assumed to be different from the color used for activation).
+	 * @returns Optimal item slot for activation under the specified constraints, or null if not found.
+	 */	
+	findStrongestSocketBonus(gear: Gear, color: GemColor, singleOnly: boolean, blacklistedColor: GemColor | null): ItemSlot | null {
+		let optimalSlot: ItemSlot | null = null;
+		let maxSocketBonusEP: number = 1e-8;
+
+		for (var slot of gear.getItemSlots()) {
+			const item = gear.getEquippedItem(slot);
+
+			if (!item) {
+				continue;
+			}
+
+			if (item!.numSocketsOfColor(blacklistedColor) != 0) {
+				continue;
+			}
+
+			const numSockets = item!.numSocketsOfColor(color);
+
+			if ((numSockets == 0) || (singleOnly && (numSockets != 1))) {
+				continue;
+			}
+
+			const socketBonusEP = new Stats(item.item.socketBonus).computeEP(this.epWeights);
+			const normalizedEP = socketBonusEP / numSockets;
+
+			if (normalizedEP > maxSocketBonusEP) {
+				optimalSlot = slot;
+				maxSocketBonusEP = normalizedEP;
+			}
+		}
+
+		return optimalSlot;
+	}
+	
+	socketGemInFirstMatchingSocket(gear: Gear, itemSlot: ItemSlot | null, colorToMatch: GemColor, gemId: number): Gear {
+		if (itemSlot != null) {
+			const item = gear.getEquippedItem(itemSlot);
+
+			for (const [socketIdx, socketColor] of item!.allSocketColors().entries()) {
```

Comment:
instead of item! do we want do just return early if `!item` on line 149-150?

Reply:
My initial assumption was that `item` would always be non-null if `itemSlot` is non-null, since `findStrongestSocketBonus()` already has a null check in it. But I added the early return anyway just for completeness.